File: js-compressor/file_type.js
const FileType = {
    TEXT: 'text',
    BINARY: 'binary',
    IMAGE: 'image',
    UNKNOWN: 'unknown'
};

const Algorithm = {
    RLE: 'rle',
    LZ: 'lz'
};

function detectFileType(data) {
    // Check if it's text (ASCII/UTF-8)
    const isText = data.every(byte => 
        (byte >= 32 && byte <= 126) || // Printable ASCII
        byte === 10 || // Newline
        byte === 13 || // Carriage return
        byte === 9     // Tab
    );
    if (isText) {
        return FileType.TEXT;
    }

    // Check for image file signatures
    if (data.length >= 8) {
        const header = data.slice(0, 8);
        if (header[0] === 0x89 && header[1] === 0x50 && header[2] === 0x4E && header[3] === 0x47 || // PNG
            header[0] === 0xFF && header[1] === 0xD8 && header[2] === 0xFF || // JPEG
            (header[0] === 0x47 && header[1] === 0x49 && header[2] === 0x46 && header[3] === 0x38 && // GIF
             (header[4] === 0x37 || header[4] === 0x39) && header[5] === 0x61)) {
            return FileType.IMAGE;
        }
    }

    return FileType.BINARY;
}

function calculateRleRatio(data) {
    if (data.length === 0) return 0;

    let currentByte = data[0];
    let currentCount = 1;
    let totalRuns = 1;
    let totalBytes = 1;

    for (let i = 1; i < data.length; i++) {
        if (data[i] === currentByte) {
            currentCount++;
        } else {
            currentByte = data[i];
            currentCount = 1;
            totalRuns++;
        }
        totalBytes++;
    }

    return (totalRuns * 2) / totalBytes;
}

function selectAlgorithm(fileType, data) {
    switch (fileType) {
        case FileType.TEXT:
            return Algorithm.LZ;
        case FileType.BINARY:
            const rleRatio = calculateRleRatio(data);
            return rleRatio > 0.5 ? Algorithm.RLE : Algorithm.LZ;
        case FileType.IMAGE:
            return Algorithm.RLE;
        default:
            return Algorithm.LZ;
    }
}

module.exports = {
    FileType,
    Algorithm,
    detectFileType,
    selectAlgorithm
}; 

File: rs-compressor/src/algos/file_type.rs
use std::collections::HashMap;

#[derive(Debug, PartialEq)]
pub enum FileType {
    Text,
    Binary,
    Image,
    Unknown,
}

#[derive(Debug, PartialEq)]
pub enum Algorithm {
    RLE,
    LZ,
}

pub fn detect_file_type(data: &[u8]) -> FileType {
    // Check for common text file characteristics
    let is_text = data.iter().all(|&b| b.is_ascii() || b == b'\n' || b == b'\r' || b == b'\t');
    if is_text {
        return FileType::Text;
    }

    // Check for common image file signatures
    if data.len() >= 8 {
        let header = &data[0..8];
        if header.starts_with(b"\x89PNG\r\n\x1a\n") || 
           header.starts_with(b"\xFF\xD8\xFF") || // JPEG
           header.starts_with(b"GIF87a") || 
           header.starts_with(b"GIF89a") {
            return FileType::Image;
        }
    }

    // If not text or image, consider it binary
    FileType::Binary
}

pub fn select_algorithm(file_type: FileType, data: &[u8]) -> Algorithm {
    match file_type {
        FileType::Text => {
            // For text files, use LZ as it's better for patterns
            Algorithm::LZ
        },
        FileType::Binary => {
            // For binary files, check if RLE would be more efficient
            let rle_ratio = calculate_rle_ratio(data);
            if rle_ratio > 0.5 {
                Algorithm::RLE
            } else {
                Algorithm::LZ
            }
        },
        FileType::Image => {
            // For images, use RLE as they often have repeated pixels
            Algorithm::RLE
        },
        FileType::Unknown => {
            // Default to LZ for unknown types
            Algorithm::LZ
        }
    }
}

fn calculate_rle_ratio(data: &[u8]) -> f64 {
    if data.is_empty() {
        return 0.0;
    }

    let mut current_byte = data[0];
    let mut current_count = 1;
    let mut total_runs = 1;
    let mut total_bytes = 1;

    for &byte in &data[1..] {
        if byte == current_byte {
            current_count += 1;
        } else {
            current_byte = byte;
            current_count = 1;
            total_runs += 1;
        }
        total_bytes += 1;
    }

    // Calculate compression ratio (lower is better)
    (total_runs * 2) as f64 / total_bytes as f64
} 